### PR TITLE
[CINN] Reconstruct shape_analysis

### PR DESCRIPTION
**Pull request by Hongqing-work**
### PR Category

CINN

### PR Types

Not User Facing

### Description

Pcard-67164
This PR reconstructs the shape_analysis class to address the interface invocation risks brought by the lazy update mechanism introduced in [PR63367](https://github.com/PaddlePaddle/Paddle/pull/63367). Specifically, during the global graph shape inference, the parameters passed to the infersymbolic function have been changed to the InferSymbolicShapeContext member variable within the shape_analysis class. This ensures that only the non-lazy mode get function is called in all inference scenarios, and the lazy mode get function is activated only when used in the pass. Additionally, the reconstruction of the shape_analysis class exposes more shape symbolic inference, and the PR also completes the following fixes:
- Fixes the issue where the derived value during backward does not have a defining op, in which case a static symbol will be used.
- Fixes the InferSymbolicShape functions for concat and gather.
- Fixes the hashing issue in value_to_shape_or_data by using the id of the value as the key.

It is worth noting that this PR is still an intermediate state within the new symbolic inference update mechanism. Future work includes:
- Completely transforming the InferSymbolicShapeContext into a private member, only making the global graph inference function a friend, and removing all redundant inference function calls in other passes.

### CODE REVIEW

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h [hunk 1]
```diff
@@ -93,23 +130,20 @@ class IR_API ShapeConstraintIRAnalysis final
...
-      value_to_shape_or_data_;
-
-  symbol::ConstraintsManager constraints_manager_;
+  void InferShapeOrDataForValue(Value val);
 
-  using DimExprSubstitutionPattern =
-      std::unordered_map<symbol::DimExpr, symbol::DimExpr>;
-  DimExprSubstitutionPattern substitution_pattern_;
+ private:
+  ModuleOp m_;
```
Comment (zyfncg) [inline paddle/pir/include/dialect/shape/utils/shape_analysis.h]: `ModuleOp m_`没有使用的地方了，可以删掉

File: paddle/pir/src/dialect/shape/utils/shape_analysis.cc [hunk 2]
```diff
@@ -38,22 +39,156 @@ void ShapeConstraintIRAnalysis::Init() {
...
+  if (!HasShapeOrDataForValue(val)) {
+    PADDLE_THROW(phi::errors::Fatal(
+        "Fail to GetShapeOrDataForValue on InferSymbolicShape!"));
+  }
+
+  return value_to_shape_or_data_.at(val);
+}
+
+void InferSymbolicShapeContext::SetStaticShapeForValue(Value val) {
+  auto type_info = val.type().dyn_cast<pir::DenseTensorType>();
```
Comment (zyfncg) [inline paddle/pir/src/dialect/shape/utils/shape_analysis.cc]: 这里对Value加一些强制检查，有问题及早抛出。

File: paddle/pir/src/dialect/shape/utils/shape_analysis.cc [hunk 3]
```diff
@@ -63,8 +213,12 @@ void ShapeConstraintIRAnalysis::InferShapeOrDataForValue(Value val) {
     subgraph_ops.insert(op);
     bool has_prev_op = false;
     for (auto& operand : op->operands_source()) {
-      if (operand.impl() && !HasShapeOrDataForValue(operand)) {
-        has_prev_op = true;
+      if (operand.impl() && !context_.HasShapeOrDataForValue(operand)) {
+        if (!operand.defining_op()) {
+          SetStaticShapeForValue(operand);
```
Comment (zyfncg) [inline paddle/pir/src/dialect/shape/utils/shape_analysis.cc:264]: 这里可以先判断下operand是不是静态shape, if嵌套比较深的话可以改成if-continue的形式

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h (hunk 1 above)
Comment (Hongqing-work) [reply paddle/pir/include/dialect/shape/utils/shape_analysis.h]: Done

File: paddle/pir/src/dialect/shape/utils/shape_analysis.cc (hunk 2 above)
Comment (Hongqing-work) [reply paddle/pir/src/dialect/shape/utils/shape_analysis.cc]: Done

File: paddle/pir/src/dialect/shape/utils/shape_analysis.cc (hunk 3 above)
Comment (Hongqing-work) [reply paddle/pir/src/dialect/shape/utils/shape_analysis.cc:264]: Done，检查直接在SetStaticShapeForValue函数里了

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h [hunk 4]
```diff
@@ -59,6 +53,49 @@ class IR_API ShapeConstraintIRAnalysis final
...
+
+  symbol::ConstraintsManager constraints_manager_;
+
+  using DimExprSubstitutionPattern =
+      std::unordered_map<symbol::DimExpr, symbol::DimExpr>;
+  DimExprSubstitutionPattern substitution_pattern_;
+};
+
+// The implementation is based on shape constraint ir.
+// The implementation is based on shape constraint ir.
```
Comment (zyfncg) [inline paddle/pir/include/dialect/shape/utils/shape_analysis.h]: 重复注释，现在这句注释没有价值，可以在下个PR里重新写一下

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h [hunk 5]
```diff
@@ -129,6 +162,8 @@ class IR_API ShapeAnalysisManager {
 
 #define OP_DECLARE_INFER_SYMBOLIC_SHAPE(name) \
   bool name##OpInferSymbolicShape(            \
-      pir::Operation* op, pir::ShapeConstraintIRAnalysis* shape_analysis);
+      pir::Operation* op, pir::InferSymbolicShapeContext* infer_context);
+
+bool IsStaticShape(const Value& value);
```
Comment (zyfncg) [inline paddle/pir/include/dialect/shape/utils/shape_analysis.h:165]: 这个函数声明放在.cc文件里吧，加到头文件影响范围会扩大

**Review by zyfncg — DISMISSED** [review]

**Review by tc20042008 — DISMISSED** [review]

**Review by zyfncg — APPROVED** [review]

**Review by tc20042008 — APPROVED** [review]

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h (hunk 4 above)
Comment (Hongqing-work) [reply paddle/pir/include/dialect/shape/utils/shape_analysis.h]: Done

File: paddle/pir/include/dialect/shape/utils/shape_analysis.h (hunk 5 above)
Comment (Hongqing-work) [reply paddle/pir/include/dialect/shape/utils/shape_analysis.h:165]: 已增加VectorType的支持，作为公共函数以供其它地方使用

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by zhangbo9674 — APPROVED** [review]

**Pull request merged**